homeassistant/components/automation/sun.py

```diff
@@ -67,6 +54,70 @@ def trigger(hass, config, action):
     return True
 
 
+def if_action(hass, config):
+    """ Wraps action method with sun based condition. """
+    before = config.get(CONF_BEFORE)
+    after = config.get(CONF_AFTER)
+
+    # Make sure required configuration keys are present
+    if before is None and after is None:
+        logging.getLogger(__name__).error(
+            "Missing if-condition configuration key %s or %s",
+            CONF_BEFORE, CONF_AFTER)
+        return None
+
+    # Make sure configuration keys have the right value
+    if before is not None and before not in (EVENT_SUNRISE, EVENT_SUNSET) or \
+       after is not None and after not in (EVENT_SUNRISE, EVENT_SUNSET):
+        logging.getLogger(__name__).error(
+            "%s and %s can only be set to %s or %s",
+            CONF_BEFORE, CONF_AFTER, EVENT_SUNRISE, EVENT_SUNSET)
+        return None
+
+    before_offset = _parse_offset(config.get(CONF_BEFORE_OFFSET))
+    after_offset = _parse_offset(config.get(CONF_AFTER_OFFSET))
+    if before_offset is False or after_offset is False:
+        return None
+
+    if before is None:
+        before_func = lambda: None
+    elif before == EVENT_SUNRISE:
+        before_func = lambda: sun.next_rising_utc(hass) + before_offset
+    else:
+        before_func = lambda: sun.next_setting_utc(hass) + before_offset
+
+    if after is None:
+        after_func = lambda: None
+    elif after == EVENT_SUNRISE:
+        after_func = lambda: sun.next_rising_utc(hass) + after_offset
+    else:
+        after_func = lambda: sun.next_setting_utc(hass) + after_offset
+
+    # This is needed for testing
```

Comment:
You're referring to 

```
    # This is needed for testing
    time_func = dt_util.utcnow
```

right?

If you can find a better way to do this, that would be great. I couldn't figure out how to patch the internal function (`time_if`) so that a call to `dt_util.utcnow` could be set in the tests.